common/modules/gcp-project/main.tf

```diff
@@ -0,0 +1,89 @@
+# this Terraform module creates the same resources as project_init script and
+# set the DNS zones structure.
+
+terraform {
+  backend "gcs" {}
+}
+
+variable "project_name" {}
+variable "project_owner" {}
+variable "billing_id" {}
+variable "organization_id" {}
+
+# compute the project_id to get the dns zone in JSON format
+# I wasn't able to replicate this function using HCL code
+data "external" "calculate_dns_zone" {
+  program = ["python", "-c", "print(\"{\\\"zone\\\": \\\"\" + \".\".join(\"${google_project.project.project_id}\".split(\"-\")[::-1]) + \".net.\") + \"\\\"}\""]
```

Comment:
```
$ python -c 'print("{\"zone\": \"" + ".".join("gpii-gcp-dev-alfredo".split("-")[::-1]) + ".net." + "\"}")'
{"zone": "alfredo.dev.gcp.gpii.net."}
```
it takes the name of the project `gpii-gcp-dev-alfredo` and returns the DNS zone in json format: `{"zone": "alfredo.dev.gcp.gpii.net."}`

The command split the project name by the hyphens and rejoin them in the other way around ([::-1]), adds a .net at the end and wrap all in {"zone": "dnsname"}.

I don't know how to do this in bash, I just wanted to reduce the parameters of the module. Perhaps using something like [this terraform function](https://github.com/hashicorp/terraform/issues/9404) but it's not implemented yet.
I used python because Exekube containers [have it installed](https://github.com/exekube/exekube/blob/master/dockerfiles/google/Dockerfile#L34)

Perhaps I can expand the comment about this resource in the .tf in order to understand what it does.

Reply:
@amatas This works:
```
variable "project_name" {
  default = "gpii-gcp-dev-alfredo"
}

output "project_zone" {
  value = "{\"zone\": \"${replace(var.project_name, "/([a-z]+)-([a-z]+)-([a-z]+)-([a-z]+)/", "$4.$3.$2.$1")}.\"}"
}
```

```
Apply complete! Resources: 4 added, 0 changed, 0 destroyed.

Outputs:

project_zone = {"zone": "alfredo.dev.gcp.gpii."}
```

Reply:
The possible input values are:
```
gpii-gcp-dev
gpii-gcp-dev-alfredo
```
this interpolation doesn't work with `gpii-gcp-dev` because it the length of the items is fixed.
```
variable "project_name" {
  default = "gpii-gcp-dev"
}

Apply complete! Resources: 0 added, 0 changed, 0 destroyed.

Outputs:

project_zone = {"zone": "gpii-gcp-dev."}
```

Reply:
@amatas This works for both variants:
```
variable "project_name" {
  default = "gpii-gcp-dev"
}

output "project_zone" {
  value = "{\"zone\": \"${
    replace(
      replace(
        var.project_name,
        "/([a-z]+)-([a-z]+)-([a-z]+)-?([a-z]+)?/",
        "$4.$3.$2.$1."),
      "/^\\./",
      "")
    }\"}"
}
```
Not sure if it is more readable than your solution, but at least it does not have any python in it :)

Reply:
I think Sergey's version is about as readable as the Python one-liner. (The part I found trickiest was removing the leading `.` when `$4` is undefined. :)) I would write a comment explaining the general problem (and maybe the leading dot thing) and adopt this solution.